src/components/vm/overview/vsock.jsx

```diff
@@ -0,0 +1,259 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import cockpit from 'cockpit';
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { Flex, FlexItem } from "@patternfly/react-core/dist/esm/layouts/Flex/index.js";
+import { Form, FormGroup } from "@patternfly/react-core/dist/esm/components/Form/index.js";
+import { NumberInput } from "@patternfly/react-core/dist/esm/components/NumberInput/index.js";
+import { Modal } from "@patternfly/react-core/dist/esm/components/Modal/index.js";
+import { Radio } from "@patternfly/react-core/dist/esm/components/Radio/index.js";
+import { useDialogs } from 'dialogs.jsx';
+
+import { ModalError } from 'cockpit-components-inline-notification.jsx';
+import { domainRemoveVsock, domainSetVsock } from "../../../libvirtApi/domain.js";
+import { WarningInactiveAlert, WarningInactiveTooltip } from "../../common/warningInactive.jsx";
+
+const _ = cockpit.gettext;
+
+const ASSIGN_AUTOMATICALLY = _("Assign automatically");
+const SET_MANUALLY = _("Set manually");
+// There  are  several  reserved  addresses:
+// VMADDR_CID_ANY  (-1U)  means  any  address  for  binding;
+// VMADDR_CID_HYPERVISOR (0) is reserved for services built into the hypervisor;
+// VMADDR_CID_LOCAL (1) is the well-known address for local communication (loopback);
+// VMADDR_CID_HOST (2) is the well-known address of the host.
+const MIN_VSOCK_CID = 3;
+
+function getVsockUsageMessage(vmName, connectionName, vms, auto, address) {
+    if (auto)
+        return;
+
+    const vmsUsingCIDAddress = vms
+            .filter(vm => !(vmName === vm.name && connectionName === vm.connectionName) && vm.vsock.cid.auto === "no" && vm.vsock.cid.address === String(address))
+            .map(vm => vm.name);
+    if (vmsUsingCIDAddress.length === 0)
+        return;
+
+    const vmsUsingCIDAddressString = vmsUsingCIDAddress.join(", ");
+
+    return cockpit.format(_("This context identifier is already used by $0. Assigning the same CID is possible, but the VMs won't be able to run simultaneously."), vmsUsingCIDAddressString);
```

Comment:
What do you propose? To not be able to assign a CID to offline VM if some other VM also has the same CID? That's not what qemu, libvirt does, and I think we should not introduce this kind of check either.

In fact, I think we should not do any checks against non-running VMs in general.
Non-running VM is basically just a configuration template. You may want to migrate them to another host. Or some other VM might have been migrated to your host, and then you might end up with several VMs with the same CID anyway.  Or you may want one VM as backup of another, or as part of VM cluster. E.g. have one VM running, and another VM acting a backup of first VM if it fails, and thus they have to be configured the same way, and have same CID (even if only one is running at once).

A while back, were solving the same issue with disks, where we implemented disk sharing between multiple VMs for clustering purposes, which however cannot run at the same time. But they have same configuration, have disks attached, so if one VM fails, another can act as backup and access the same disk. Now, I don't know how many (if any) people would use vsocks in their cluster setups though, so maybe it's just a corner case.

But in summary, I think we should not prevent user from assigning the same CIDs to multiple VMs. We should warn them, but not prevent them.

Reply:
I wasn't suggesting to not allow it, I was just curious about the use case -- in particular, why users would want to specify them manually and therefore introduce potential conflicts. Backup VMs seem moderately plausible.

> In fact, I think we should not do any checks against non-running VMs in general. 

The PR already does that, no? It needs to scan used CIDs for conflicts. I don't object, just pointing it out.

Reply:
> The PR already does that, no? It needs to scan used CIDs for conflicts. I don't object, just pointing it out.

I meant checks where we would then prevent users from doing some kind of action. We should throw a warning, but never prevent some action because of a theoretical conflict with some shutoff VM.